routing/pathfind_test.go

```diff
@@ -291,18 +291,42 @@ type expectedHop struct {
 
 type basicGraphPathFindingTestCase struct {
 	target        string
+	paymentAmt    btcutil.Amount
 	totalTimeLock uint32
 	expectedHops  []expectedHop
 }
 
 var basicGraphPathFindingTests = []basicGraphPathFindingTestCase{
-	{target: "sophon", totalTimeLock: 102, expectedHops: []expectedHop{
+	// Basic route with one intermediate hop
+	{ target: "sophon", paymentAmt: 100, totalTimeLock: 102, 
+	  expectedHops: []expectedHop{
 		{alias: "songoku", fwdAmount: 100000, fee: 110, timeLock: 101},
 		{alias: "sophon", fwdAmount: 100000, fee: 0, timeLock: 101},
 	}},
-	{target: "luoji", totalTimeLock: 101, expectedHops: []expectedHop{
+	// Basic direct (one hop) route
+	{ target: "luoji", paymentAmt: 100, totalTimeLock: 101, 
+	  expectedHops: []expectedHop{
 		{alias: "luoji", fwdAmount: 100000, fee: 0, timeLock: 101},
-	}}}
+	}},
+	// Three hop route where fees need to be added in to the forwarding amount.
+	// The high fee hop phamnewun should be avoided
+	{ target: "elst", paymentAmt: 50000, totalTimeLock: 103, 
+ 	  expectedHops: []expectedHop{
+		{alias: "songoku", fwdAmount: 50000200, fee: 50010, timeLock: 102},
+		{alias: "sophon", fwdAmount: 50000000, fee: 200, timeLock: 101},
+		{alias: "elst", fwdAmount: 50000000, fee: 0, timeLock: 101},
+	}},
+	// Three hop route where fees need to be added in to the forwarding amount.
+	// However this time the fwdAmount becomes too large for the roasbeef <->
+	// songoku channel. Then there is no other option than to choose the
+	// expensive phamnuwen channel. This test case was failing before
+	// the route search was executed backwards.
+	{ target: "elst", paymentAmt: 100000, totalTimeLock: 103, 
+ 	  expectedHops: []expectedHop{
+	       {alias: "phamnuwen", fwdAmount: 100000200, fee: 100010000, timeLock: 102},
```

Comment:
I don't get this to add up, maybe I'm missing something. 

For a payment of `100,000,000 msat`, the expected fees should be:
sophon->elst: `200 msat + 100,000,000 msat * 0/1,000,000 msat = 200 msat`
pham->sophon: `10,000 msat + 100,000,200 msat * 1,000,000/1,000,000 msat = 100,010,200 msat`

We must send `paymentAmt+fees = 100,000,000 msat + 200 msat + 100,010,200 msat = 200,010,400 msat`. This should not be able to go through the `120,000,000 msat` channel?

Reply:
Great catch! This is indeed not correct. I hope that conversion to table-driven made this easier to spot. And I think it actually uncovers two existing bugs in newRoute on master:

- Fee calculation for a hop does not include the fee that needs to be paid to the next hop
- The incoming channel capacity "sanity" check does not include the fee to be paid to the current hop

I also made the expected total amount for a route explicit in the test table. Need to be careful with calculating expected values in unit tests, to prevent duplication of the same (incorrect logic). In this case, the calculation of the expected value was ok, but still.

See c21c9802a7c56c667c9a3deac3bf76dee704a171 for fixes.